taichi/system/timer.cpp

```diff
@@ -69,6 +69,24 @@ void Time::sleep(double s) {
   Time::usleep(s * 1e6_f64);
 }
```

Comment:
Here is a mush more accurate milliseconds level sleep implementation instead of ```Sleep(DWORD(us * 1e-3))``` on windows . 
```
#ifdef _WIN64

#include <Windows.h>
#pragma comment(lib, "Winmm.lib")

void win_sleep(DWORD ms) {
  if (ms == 0)
    Sleep(0);
  else {
    HANDLE hEvent = CreateEvent(NULL, FALSE, FALSE, NULL);
    timeSetEvent(ms, 1, (LPTIMECALLBACK)hEvent , 0, TIME_ONESHOT | TIME_CALLBACK_EVENT_SET);
    WaitForSingleObject(hEvent, INFINITE);
    CloseHandle(hEvent);
  }
}
#endif
```

Reply:
Sounds cool! But is that possible to sleep for microseconds (us) on Windows?

Reply:
It's sad truth that sleep accuracy can not be guaranteed when delay time is lower than 1ms :(
The only possible approach is to use high resolution clock to continually counting in a for loop , but it 's unable to release CPU time slice ...

Reply:
> Sounds cool! But is that possible to sleep for microseconds (us) on Windows?

It's sad truth that sleep accuracy can not be guaranteed when delay time is lower than 1ms :(
The only possible approach is to use high resolution clock to continually counting in a for loop , but it 's unable to release CPU time slice ...

Reply:
Related problem : [is-there-a-windows-equivalent-of-nanosleep](https://stackoverflow.com/questions/7827062/is-there-a-windows-equivalent-of-nanosleep)

Reply:
Sleeping for us seems tricky on win, how about just:
```cpp
#ifndef _WIN64
} while (dt > 1e-4_f64);  // until dt <= 100us
#else
} while (dt > 1e-2_f64);  // until dt <= 10ms
#endif
```

Reply:
1ms is more accurate ( 1/60 = 16.6 ms  , 1/120 = 8.3ms , 1/40 = 25 ms )
···
#else
} while (dt > 1e-1_f64);  // until dt <= 1ms
#endif
···